File: dev-utils-service/src/main/java/com/ttg/devknowledgeplatform/devutils/api/DevUtilsApi.java
package com.ttg.devknowledgeplatform.devutils.api;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;

import com.ttg.devknowledgeplatform.devutils.dto.DevUtilResponse;
import com.ttg.devknowledgeplatform.devutils.dto.MinifiableTextRequest;
import com.ttg.devknowledgeplatform.devutils.dto.StringCaseResponse;
import com.ttg.devknowledgeplatform.devutils.dto.TextRequest;

import jakarta.validation.Valid;

/**
 * HTTP contract for the stateless developer-utility operations. The implementation
 * ({@link com.ttg.devknowledgeplatform.devutils.api.impl.DevUtilsController}) carries no HTTP
 * annotations, same split as every other module's {@code api}/{@code api.impl} pair.
 *
 * <p>Every endpoint here is public — no {@code @CurrentUserId}, no authenticated principal at all.
 * See {@code security.SecurityConfig}'s Javadoc and root {@code CLAUDE.md}'s Security section for
 * why this is the one deployable in the reactor that isn't a JWT resource server, and this
 * module's own {@code CLAUDE.md} for the matching carve-out {@code gateway}'s own routing needs.
 *
 * <p>Each endpoint takes whichever request DTO actually fits its own operation — most genuinely
 * share {@link MinifiableTextRequest}'s shape (raw text in, a minify flag, transformed text out);
 * {@code jsonToYaml}/{@code jsonToCsv} do not, since neither YAML nor CSV has a distinct "compact"
 * form to toggle (see {@link TextRequest}'s own Javadoc) — rather than every endpoint being forced
 * through one shared request type; see {@code service.DevUtilOperation}'s own Javadoc for the full
 * reasoning.
 */
@RequestMapping("/api/v1/dev-utils")
public interface DevUtilsApi {

    /**
     * Validates a raw JSON string and returns it pretty-printed, or, with
     * {@code request.minify()}, compact/single-line.
     *
     * @return {@code 200} with the formatted JSON, or {@code 400} if {@code request.input()} is
     *         not valid JSON
     */
    @PostMapping("/json/format")
    ResponseEntity<DevUtilResponse> formatJson(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts a raw YAML string to JSON, pretty-printed or (with {@code request.minify()})
     * compact/single-line.
     *
     * @return {@code 200} with the converted JSON, or {@code 400} if {@code request.input()} is
     *         not valid YAML
     */
    @PostMapping("/yaml-to-json")
    ResponseEntity<DevUtilResponse> yamlToJson(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts a raw JSON string to YAML. No minify option — see {@link TextRequest}'s own
     * Javadoc for why.
     *
     * @return {@code 200} with the converted YAML, or {@code 400} if {@code request.input()} is
     *         not valid JSON
     */
    @PostMapping("/json-to-yaml")
    ResponseEntity<DevUtilResponse> jsonToYaml(@Valid @RequestBody TextRequest request);

    /**
     * Reformats raw HTML with consistent indentation, or, with {@code request.minify()}, jsoup's
     * own unformatted output mode.
     *
     * @return {@code 200} with the beautified (or minified) HTML
     */
    @PostMapping("/html/beautify")
    ResponseEntity<DevUtilResponse> beautifyHtml(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats raw CSS with consistent indentation, or, with {@code request.minify()}, a compact
     * form with comments and non-essential whitespace stripped. Lenient — see
     * {@code CssOperation}'s own Javadoc — this never fails on malformed input.
     *
     * @return {@code 200} with the beautified (or minified) CSS
     */
    @PostMapping("/css/beautify")
    ResponseEntity<DevUtilResponse> beautifyCss(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats raw LESS with consistent indentation, or, with {@code request.minify()}, a compact
     * form with comments and non-essential whitespace stripped. Reformats only — does not compile
     * LESS to plain CSS (see {@code LessOperation}'s own Javadoc) — and, like {@code beautifyCss},
     * never fails on malformed input.
     *
     * @return {@code 200} with the beautified (or minified) LESS
     */
    @PostMapping("/less/beautify")
    ResponseEntity<DevUtilResponse> beautifyLess(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats raw SCSS with consistent indentation, or, with {@code request.minify()}, a compact
     * form with comments and non-essential whitespace stripped. Reformats only — does not compile
     * SCSS to plain CSS (see {@code ScssOperation}'s own Javadoc) — and, like {@code beautifyCss},
     * never fails on malformed input.
     *
     * @return {@code 200} with the beautified (or minified) SCSS
     */
    @PostMapping("/scss/beautify")
    ResponseEntity<DevUtilResponse> beautifyScss(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats raw JavaScript with consistent indentation, or, with {@code request.minify()}, a
     * compact form with comments and non-essential whitespace stripped. See {@code JsOperation}'s
     * own Javadoc for its Automatic Semicolon Insertion (ASI) safety guarantee — this never fails
     * on malformed input.
     *
     * @return {@code 200} with the beautified (or minified) JavaScript
     */
    @PostMapping("/js/beautify")
    ResponseEntity<DevUtilResponse> beautifyJs(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats ERB (Embedded RuBy) markup with consistent indentation, or, with
     * {@code request.minify()}, jsoup's own unformatted output mode — {@code <% %>}/{@code <%= %>}
     * tags are preserved verbatim regardless (see {@code ErbOperation}'s own Javadoc). Never fails
     * on malformed input, the same jsoup-backed leniency {@code beautifyHtml} has.
     *
     * @return {@code 200} with the beautified (or minified) ERB
     */
    @PostMapping("/erb/beautify")
    ResponseEntity<DevUtilResponse> beautifyErb(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Validates a raw XML string and re-serializes it indented, or, with {@code request.minify()},
     * with insignificant inter-element whitespace stripped.
     *
     * @return {@code 200} with the beautified (or minified) XML, or {@code 400} if
     *         {@code request.input()} is not well-formed XML
     */
    @PostMapping("/xml/beautify")
    ResponseEntity<DevUtilResponse> beautifyXml(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts a raw JSON array of objects (or a single object) to CSV. No minify option — see
     * {@link TextRequest}'s own Javadoc for why; CSV has no distinct "compact" form to toggle.
     *
     * @return {@code 200} with the converted CSV, or {@code 400} if {@code request.input()} isn't
     *         valid JSON shaped as an array of objects (or a single object)
     */
    @PostMapping("/json-to-csv")
    ResponseEntity<DevUtilResponse> jsonToCsv(@Valid @RequestBody TextRequest request);

    /**
     * Converts raw CSV (first row treated as the header) to a JSON array of objects, pretty-printed
     * or (with {@code request.minify()}) compact/single-line.
     *
     * @return {@code 200} with the converted JSON, or {@code 400} if {@code request.input()} is
     *         not valid CSV
     */
    @PostMapping("/csv-to-json")
    ResponseEntity<DevUtilResponse> csvToJson(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Reformats a raw SQL query/script with one clause per line, or, with
     * {@code request.minify()}, a compact form with comments stripped. Lenient — see
     * {@code SqlFormatOperation}'s own Javadoc — this never fails on malformed input.
     *
     * @return {@code 200} with the formatted (or minified) SQL
     */
    @PostMapping("/sql/format")
    ResponseEntity<DevUtilResponse> formatSql(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts a raw PHP array literal ({@code [...]} or legacy {@code array(...)}) to JSON,
     * pretty-printed or (with {@code request.minify()}) compact/single-line.
     *
     * @return {@code 200} with the converted JSON, or {@code 400} if {@code request.input()} is
     *         not a valid PHP array literal
     */
    @PostMapping("/php-to-json")
    ResponseEntity<DevUtilResponse> phpToJson(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts a raw JSON value to a PHP array literal, one clause per line, or (with
     * {@code request.minify()}) compact/single-line.
     *
     * @return {@code 200} with the converted PHP, or {@code 400} if {@code request.input()} is
     *         not valid JSON
     */
    @PostMapping("/json-to-php")
    ResponseEntity<DevUtilResponse> jsonToPhp(@Valid @RequestBody MinifiableTextRequest request);

    /**
     * Converts raw text into every {@link StringCaseResponse} case variant at once (camelCase,
     * PascalCase, snake_case, kebab-case, CONSTANT_CASE, Title Case, Sentence case). No minify
     * option — there's no "compact form" of a case conversion — and this never fails on any input.
     *
     * @return {@code 200} with every case variant
     */
    @PostMapping("/string-case/convert")
    ResponseEntity<StringCaseResponse> convertStringCase(@Valid @RequestBody TextRequest request);

    /**
     * Encodes raw text as a standard Base64 string. No minify option — see {@link TextRequest}'s
     * own Javadoc for why; a Base64 encoding has no distinct "compact form" to toggle. Never fails
     * — every string has a valid encoding.
     *
     * @return {@code 200} with the Base64-encoded text
     */
    @PostMapping("/base64/encode")
    ResponseEntity<DevUtilResponse> encodeBase64(@Valid @RequestBody TextRequest request);

    /**
     * Decodes a standard Base64 string back into text.
     *
     * @return {@code 200} with the decoded text, or {@code 400} if {@code request.input()} isn't
     *         valid Base64
     */
    @PostMapping("/base64/decode")
    ResponseEntity<DevUtilResponse> decodeBase64(@Valid @RequestBody TextRequest request);

    /**
     * Percent-encodes raw text for safe use inside a URL. No minify option — see
     * {@link TextRequest}'s own Javadoc for why; a percent-encoding has no distinct "compact form"
     * to toggle. Never fails — every string has a valid encoding.
     *
     * @return {@code 200} with the URL-encoded text
     */
    @PostMapping("/url/encode")
    ResponseEntity<DevUtilResponse> encodeUrl(@Valid @RequestBody TextRequest request);

    /**
     * Decodes a percent-encoded URL string back into text.
     *
     * @return {@code 200} with the decoded text, or {@code 400} if {@code request.input()} isn't
     *         validly percent-encoded
     */
    @PostMapping("/url/decode")
    ResponseEntity<DevUtilResponse> decodeUrl(@Valid @RequestBody TextRequest request);

    /**
     * Escapes {@code & < > " '} into their named HTML character references. No minify option —
     * see {@link TextRequest}'s own Javadoc for why; an escaped form has no distinct "compact
     * form" to toggle. Never fails — every string has a valid escaped form.
     *
     * @return {@code 200} with the HTML-entity-encoded text
     */
    @PostMapping("/html-entity/encode")
    ResponseEntity<DevUtilResponse> encodeHtmlEntity(@Valid @RequestBody TextRequest request);

    /**
     * Decodes HTML character references back into their literal characters. Never fails — an
     * unrecognized {@code &...;} sequence is left untouched rather than rejected.
     *
     * @return {@code 200} with the decoded text
     */
    @PostMapping("/html-entity/decode")
    ResponseEntity<DevUtilResponse> decodeHtmlEntity(@Valid @RequestBody TextRequest request);
}

File: gui/src/features/dev-utils/api/devUtilsApi.ts
import { httpClient } from '@shared/api/httpClient';
import { DevUtilsResponse, HashResponse, StringCaseResponse } from '../types';

type ShowError = (message: string) => void;

// gateway (dev-utils-service is public, no auth — see that service's own SecurityConfig and
// gateway's matching permitAll() carve-out).
const BASE = '/api/v1/dev-utils';

// One method per backend operation, not a generic dispatcher, since a few genuinely have a
// different shape (jsonToYaml/jsonToCsv have no `minify` parameter at all; convertStringCase
// returns StringCaseResponse, not DevUtilsResponse) — a shared `post` helper would either have to
// paper over those differences or still need per-method wrappers anyway, so it wouldn't actually
// remove any of this file's real duplication, just the `BASE +` prefix, which the constant above
// already handles.
export const devUtilsApi = {
  formatJson(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/json/format`, { input, minify }, showError);
  },

  yamlToJson(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/yaml-to-json`, { input, minify }, showError);
  },

  // No `minify` field at all — the backend operation doesn't support it (jackson-dataformat-yaml
  // has no single-line/flow-style toggle), so the key is never sent.
  jsonToYaml(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/json-to-yaml`, { input }, showError);
  },

  beautifyHtml(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/html/beautify`, { input, minify }, showError);
  },

  // The 6 CSS/LESS/SCSS/JS/ERB/XML operations added alongside dev-utils-service's own
  // CurlyBraceFormatter/ErbOperation/XmlOperation — every one of them shares the identical
  // "text in, minify flag, text out" shape `beautifyHtml` already has, so each is just a thin
  // pass-through to its own endpoint, same as every method above.
  beautifyCss(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/css/beautify`, { input, minify }, showError);
  },

  beautifyLess(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/less/beautify`, { input, minify }, showError);
  },

  beautifyScss(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/scss/beautify`, { input, minify }, showError);
  },

  beautifyJs(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/js/beautify`, { input, minify }, showError);
  },

  beautifyErb(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/erb/beautify`, { input, minify }, showError);
  },

  // The one of the 6 with a real backend validation/error path (dev-utils-service's XmlOperation
  // is JAXP-backed, unlike the lenient CurlyBraceFormatter-based CSS/LESS/SCSS/JS ones or the
  // jsoup-based ERB one) — same shape here regardless, the difference only matters to
  // errorFormatting.ts's own message handling on a failed submit.
  beautifyXml(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/xml/beautify`, { input, minify }, showError);
  },

  // JSON<->CSV + SQL Formatter, added alongside dev-utils-service's own JsonToCsvOperation/
  // CsvToJsonOperation/SqlFormatOperation. jsonToCsv has no `minify` field, same reasoning as
  // jsonToYaml above (CSV has no distinct "compact" form to toggle either).
  jsonToCsv(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/json-to-csv`, { input }, showError);
  },

  // The other operation (besides yamlToJson/beautifyXml) with a real backend invalid-input error
  // path — dev-utils-service's CsvToJsonOperation is backed by a genuine Jackson CsvMapper parser.
  csvToJson(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/csv-to-json`, { input, minify }, showError);
  },

  formatSql(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/sql/format`, { input, minify }, showError);
  },

  // PHP<->JSON + String Case Converter, added alongside dev-utils-service's own
  // PhpToJsonOperation/JsonToPhpOperation/StringCaseOperation.
  //
  // The 4th operation (besides yamlToJson/beautifyXml/csvToJson) with a real backend
  // invalid-input error path — dev-utils-service's PhpToJsonOperation is backed by a genuine
  // recursive-descent parser (PhpArrayParser).
  phpToJson(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/php-to-json`, { input, minify }, showError);
  },

  jsonToPhp(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/json-to-php`, { input, minify }, showError);
  },

  // Returns StringCaseResponse, not DevUtilsResponse — the one operation whose output is
  // genuinely richer than a single string (see that type's own comment). No `minify` field —
  // there's no "compact form" of a case conversion.
  convertStringCase(input: string, showError?: ShowError): Promise<StringCaseResponse> {
    return httpClient.post(`${BASE}/string-case/convert`, { input }, showError);
  },

  // The first ENCODERS_DECODERS-group operation, added alongside dev-utils-service's own
  // Base64EncodeOperation/Base64DecodeOperation. No `minify` field on either — a Base64 encoding
  // has no distinct "compact form" to toggle, same reasoning `jsonToYaml`/`jsonToCsv` already
  // establish. Two separate methods (not one "base64" method with a direction flag), matching how
  // every other bidirectional pair in this file (`phpToJson`/`jsonToPhp`, `yamlToJson`/
  // `jsonToYaml`) already gets one method per direction.
  encodeBase64(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/base64/encode`, { input }, showError);
  },

  decodeBase64(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/base64/decode`, { input }, showError);
  },

  // The second ENCODERS_DECODERS-group operation, added alongside dev-utils-service's own
  // UrlEncodeOperation/UrlDecodeOperation. No `minify` field on either — same reasoning
  // `encodeBase64`/`decodeBase64` already establish; a percent-encoding has no distinct "compact
  // form" to toggle.
  encodeUrl(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/url/encode`, { input }, showError);
  },

  decodeUrl(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/url/decode`, { input }, showError);
  },

  // The third ENCODERS_DECODERS-group operation, added alongside dev-utils-service's own
  // HtmlEntityEncodeOperation/HtmlEntityDecodeOperation. No `minify` field on either — same
  // reasoning `encodeBase64`/`encodeUrl` already establish; an escaped form has no distinct
  // "compact form" to toggle.
  encodeHtmlEntity(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/html-entity/encode`, { input }, showError);
  },

  decodeHtmlEntity(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/html-entity/decode`, { input }, showError);
  },

  // Added alongside dev-utils-service's own HashGeneratorOperation (originally the first
  // INSPECTORS-group operation, later moved to ENCODERS_DECODERS — see that class's own Javadoc).
  // Returns HashResponse, not DevUtilsResponse — the second operation (after convertStringCase)
  // whose output is genuinely richer than a single string. No `minify` field — a hash digest has
  // no distinct "compact form" to toggle.
  generateHash(input: string, showError?: ShowError): Promise<HashResponse> {
    return httpClient.post(`${BASE}/hash/generate`, { input }, showError);
  },

  // The fourth ENCODERS_DECODERS-group operation, added alongside dev-utils-service's own
  // PhpSerializeOperation/PhpUnserializeOperation. No `minify` field on either — same reasoning
  // `encodeBase64`/`encodeUrl`/`encodeHtmlEntity` already establish; PHP's serialize() format has
  // no distinct "compact form" to toggle, and unserializePhp's own JSON output is always
  // pretty-printed (see that backend operation's own Javadoc for why it deliberately doesn't
  // support minify the way `phpToJson` does).
  serializePhp(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/php-serialize/serialize`, { input }, showError);
  },

  unserializePhp(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/php-serialize/unserialize`, { input }, showError);
  },

  // The fifth ENCODERS_DECODERS-group operation, added alongside dev-utils-service's own
  // AsciiToHexOperation/HexToAsciiOperation. No `minify` field on either — same reasoning
  // `encodeBase64`/`encodeUrl`/`encodeHtmlEntity`/`serializePhp` already establish; a hex
  // representation has no distinct "compact form" to toggle.
  encodeHex(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/hex/encode`, { input }, showError);
  },

  decodeHex(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/hex/decode`, { input }, showError);
  },

  // The first INSPECTORS-group operation, added alongside dev-utils-service's own
  // JwtDebuggerOperation. Shares MinifiableTextRequest's own "pretty vs. minify" shape — unlike
  // every ENCODERS_DECODERS pair above, this operation's output (a JSON object) genuinely has a
  // "compact form" to toggle, the same reasoning `formatJson`/`beautifyXml` etc. already establish.
  debugJwt(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/jwt/debug`, { input, minify }, showError);
  },

  // The second INSPECTORS-group operation, added alongside dev-utils-service's own
  // RegexTesterOperation. Genuinely 3 separate fields, not `input`/`minify` — see that backend
  // operation's own `RegexTestRequest` DTO Javadoc for why this needed its own dedicated shape
  // rather than being bent into the (input, minify) convention every method above shares.
  testRegexp(pattern: string, flags: string, testText: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/regexp/test`, { pattern, flags, testText }, showError);
  },

  // The first WEB-group operation, added alongside dev-utils-service's own UrlParserOperation.
  // Back to the plain (input, minify) shape every Formatters-group operation already shares — its
  // output (protocol/hostname/port/etc., plus a parsed query object) is a single JSON string like
  // any other, so it renders through the shared DevUtilToolPanel with no bespoke panel needed.
  parseUrl(input: string, minify: boolean, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/url/parse`, { input, minify }, showError);
  },

  // The third INSPECTORS-group operation, added alongside dev-utils-service's own
  // CronParserOperation. No `minify` field — same reasoning `encodeBase64`/`generateHash`/etc.
  // already establish; a plain-English description has no distinct "compact form" to toggle.
  parseCron(input: string, showError?: ShowError): Promise<DevUtilsResponse> {
    return httpClient.post(`${BASE}/cron/parse`, { input }, showError);
  },
};
